### ISSUE TITLE
Conflict error when create pod because resource quota version is old

### ISSUE THREAD

**Issue by Nebulazhang** [kubernetes/kubernetes#67761]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
In k8s.io/client-go/util/workqueue/queue.go
[https://github.com/kubernetes/client-go/blob/master/util/workqueue/queue.go](url)
1.
```
func (q *Type) Add(item interface{}) {
	q.cond.L.Lock()
	defer q.cond.L.Unlock()
	if q.shuttingDown {
		return
	}
	if q.dirty.has(item) {
		return
	}

	q.metrics.add(item)

	q.dirty.insert(item)
	if q.processing.has(item) {
		return
	}

	q.queue = append(q.queue, item)
	q.cond.Signal()
}
```

```
func (q *Type) Get() (item interface{}, shutdown bool) {
	q.cond.L.Lock()
	defer q.cond.L.Unlock()
	for len(q.queue) == 0 && !q.shuttingDown {
		q.cond.Wait()
	}
	if len(q.queue) == 0 {
		// We must be shutting down.
		return nil, true
	}

	item, q.queue = q.queue[0], q.queue[1:]

	q.metrics.get(item)

	q.processing.insert(item)
	q.dirty.delete(item)

	return item, false
}
```
In add() function , we want to add the unique item in the  queue.
In get() function, if more than 1 consumers , they will be waiting until producer put item into the queue.

If 10  consumers are waiting and  10 same items are put into the queue, the 10 items will be all consumed.  Maybe we expect only **1** consumer to consume the same item because the item is reduplicated.
<br/>
2.
Everything is OK, but  there is some problems in kube-controller-manager.
I use the example of resource_quota_controller:
[https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/resourcequota/resource_quota_controller.go](url)
There are 6 workers are working.

```
        for i := 0; i < workers; i++ {
		go wait.Until(rq.worker(rq.queue), time.Second, stopCh)
		go wait.Until(rq.worker(rq.missingUsageQueue), time.Second, stopCh)
	}
```

```
func (rq *ResourceQuotaController) worker(queue workqueue.RateLimitingInterface) func() {
	workFunc := func() bool {
		key, quit := queue.Get()
		if quit {
			return true
		}
		defer queue.Done(key)
		err := rq.syncHandler(key.(string))
		if err == nil {
			queue.Forget(key)
			return false
		}
		utilruntime.HandleError(err)
		queue.AddRateLimited(key)
		return false
	}

	return func() {
		for {
			if quit := workFunc(); quit {
				glog.Infof("resource quota controller worker shutting down")
				return
			}
		}
	}
}
```
When we update or delete a deployment, if there are 10 pod in this deployment, there are 10 same key(the namespace/name of quota) will be put into the queue. And the 6 workers are waked up and to deal with the same key.

Quota version maybe expires after re-calculation and the expired quota will  update unsuccessfully because other workers have updated the quota.

There are too many conflict error  will be logged.
```
Operation cannot be fulfilled on resourcequotas \"pod-quota\": the object has been modified; please apply your changes to the latest version and try again
```
<br/>
3.

Up to now, there is only warning and no error is happened. 
But We maybe fail to create pod this time, the error info is : 

```
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on resourcequotas \"pod-quota\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"pod-quota","kind":"resourcequotas"},"code":409}
```
Because the resource quota will calculate and update before create a pod. And if some error occurs, the function will retry 3 times. And if api-server doesn't update the quota successfully after 3 times, it will return back an error for user.

```
e.checkQuotas(quotas, admissionAttributes, 3)
func (e *quotaEvaluator) checkQuotas(quotas []api.ResourceQuota, admissionAttributes []*admissionWaiter, remainingRetries int) {
   ......
    for i := range quotas {
		newQuota := quotas[i]

		// if this quota didn't have its status changed, skip it
		if quota.Equals(originalQuotas[i].Status.Used, newQuota.Status.Used) {
			continue
		}

		if err := e.quotaAccessor.UpdateQuotaStatus(&newQuota); err != nil {
			updatedFailedQuotas = append(updatedFailedQuotas, newQuota)
			lastErr = err
		}
   }
   e.checkQuotas(quotasToCheck, admissionAttributes, remainingRetries-1)
}
```

**How to reproduce it (as minimally and precisely as possible)**:
1. create a resource quota , and create  50 pod by 1 deployment
2. delete the deployment and you can watch so many log like this:"please apply your changes to the latest version and try again" 

**I hope that**:
1.  reduce the conflict log and error by up the log level?
2. not wait but return if the queue is empty. or add an function like "peek()"?
3. add retry delay when error occurs in create pod period?

**Comment by DahuK** [comment]
+1, met the same issue with 1.9.7 controller-manager

**Comment by wenjiaswe** [comment]
cc @jpbetz @wenjiaswe

**Comment by yue9944882** [comment]
> If 10 consumers are waiting and 10 same items are put into the queue, the 10 items will be all consumed. Maybe we expect only 1 consumer to consume the same item because the item is reduplicated.

no, the same items in the queue will be dedupped. which is just as you say, 1 consumer for 1 item exactly.

**Comment by Nebulazhang** [comment]
I'm sorry I lost these branches and PRs,  this is my first PR in K8S.

**Comment by Nebulazhang** [comment]
> > If 10 consumers are waiting and 10 same items are put into the queue, the 10 items will be all consumed. Maybe we expect only 1 consumer to consume the same item because the item is reduplicated.
> 
> no, the same items in the queue will be dedupped. which is just as you say, 1 consumer for 1 item exactly.

there are 5 worker waiting for the key at the same time. When an item arrives, it  be consumed before it is dedupped. So that maybe 5 workers update the resource quota at the same time and lead to conflict

**Comment by yue9944882** [comment]
> no, the same items in the queue will be dedupped. which is just as you say, 1 consumer for 1 item exactly.

apologies. i meant the delta fifo in the informer will dedup items on duplication, not the controller's work queue you are mentioning above.  while,  as for the work-queue in the controller, duplicated items will be saved into dirty set until its completely processed (which ends up by `Done()`) by one of the workers. so still, one resource quota will be consumed by one worker. does that make sense?

**Comment by Nebulazhang** [comment]
> > no, the same items in the queue will be dedupped. which is just as you say, 1 consumer for 1 item exactly.
> 
> apologies. i meant the delta fifo in the informer will dedup items on duplication, not the controller's work queue you are mentioning above. while, as for the work-queue in the controller, duplicated items will be saved into dirty set until its completely processed (which ends up by `Done()`) by one of the workers. so still, one resource quota will be consumed by one worker. does that make sense?

Thank you for pointing out my problems.I do make a mistake here.But my PR is correct and I will show  you my analysis

**Comment by liggitt** [comment]
to clarify, the conflict errors are seen in the quota controller log, correct? some degree of conflict is expected, so lowering the verbosity of conflict errors would be a more reasonable fix

**Comment by Nebulazhang** [comment]
> to clarify, the conflict errors are seen in the quota controller log, correct? some degree of conflict is expected, so lowering the verbosity of conflict errors would be a more reasonable fix

Yes, and the conflict errors are also seen in api-server log  when create a pod . Too many conflict errors cause pod  create to fail.

**Comment by kszalontai** [comment]
any new about this?
we're still experiencing it in master version **1.13.11-gke.5** when creating multiple jobs

**Comment by TBBle** [comment]
I believe we're seeing this on a bare-metal k8s 1.15.4 deployment when Helm v3 tries to create a dozen Services at once. The conflict state leaks back to the caller and fails the install process.

From a `helm install -v9` trace:
```
I1121 19:04:32.000471    5784 round_trippers.go:443] POST https://api.ourk8s/api/v1/namespaces/projectdemo/services 409 Conflict in 603 milliseconds
I1121 19:04:32.000471    5784 round_trippers.go:449] Response Headers:
I1121 19:04:32.001475    5784 round_trippers.go:452]     Strict-Transport-Security: max-age=15724800; includeSubDomains
I1121 19:04:32.001475    5784 round_trippers.go:452]     Server: openresty/1.15.8.1
I1121 19:04:32.002478    5784 round_trippers.go:452]     Date: Thu, 21 Nov 2019 08:04:31 GMT
I1121 19:04:32.002478    5784 round_trippers.go:452]     Content-Type: application/json
I1121 19:04:32.002478    5784 round_trippers.go:452]     Content-Length: 326
I1121 19:04:32.002478    5784 request.go:968] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on resourcequotas \"projectdemo\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"projectdemo","kind":"resourcequotas"},"code":409}
```

**Comment by ddudnik** [comment]
We are experiencing this issue, specifically with `Service` resources for some reason, although we do have resource quota on many other types of resources.

Our cluster info: 
`version.Info{
  Major:"1", Minor:"15", GitVersion:"v1.15.6", 
  GitCommit:"7015f71e75f670eb9e7ebd4b5749639d42e20079", GitTreeState:"clean", 
  BuildDate:"2019-11-13T11:11:50Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"
}
`

**Comment by TBBle** [comment]
Interestingly, we tried setting `--concurrent-resource-quota-syncs` to 1 to avoid this issue, and it apparently still happened as I posted above. We tried that based my understanding of on https://github.com/kubernetes/kubernetes/issues/67761#issuecomment-444061087 above.

**Comment by madari** [comment]
This appears to be happening again regularly (0.5% of the time in a namespace with a pod churn of ~5 pods/s): `"Operation cannot be fulfilled on resourcequotas "gke-resource-quotas": the object has been modified; please apply your changes to the latest version and try again"`.

What's not clear to me is whether this is WAI or not?

**Comment by TBBle** [comment]
@madari To try and find a common factor, does your namespace have resource quotas on the number of Services?

**Comment by TBBle** [comment]
@Nebulazhang Are you still around to reopen this issue? It aged-out, but clearly is still going on.

**Comment by madari** [comment]
Here's a snapshot of the spec and status for `gke-resource-quotas` in one of my clusters:

```
spec:
  hard:
    count/ingresses.extensions: 100k
    count/jobs.batch: 50k
    pods: 15k
    services: 5k
status:
  hard:
    count/ingresses.extensions: 100k
    count/jobs.batch: 50k
    pods: 15k
    services: 5k
  used:
    count/ingresses.extensions: "0"
    count/jobs.batch: "0"
    pods: "398"
    services: "0"
```

The only value changing is "used pods". That value changes continuously (due to high pod churn).

**Comment by aandryashin** [comment]
We are also facing the same issue in our production cluster. I was able to reproduce it even on local minikube as described above, with a resourcequota and a deployment with 50 replicas. When deployment is deleting there are many errors in kube-controller-manager-minikube pod logs:
```
E0326 14:15:22.541986       1 resource_quota_controller.go:251] Operation cannot be fulfilled on resourcequotas "quota": the object has been modified; please apply your changes to the latest version and try again
E0326 14:15:57.295237       1 resource_quota_controller.go:251] Operation cannot be fulfilled on resourcequotas "quota": the object has been modified; please apply your changes to the latest version and try again
```
```
$ kubectl version --short
Client Version: v1.16.3
Server Version: v1.16.3
```
Is there any workaround?

**Comment by TBBle** [comment]
@aandryashin To be clear, doe the failures cause the delete to fail? The error messages are cosmetic, it's only a problem if they leak back through into the API operation.

**Comment by bryanlarsen** [comment]
@liggit are you a collaborator?  Can you reopen please?

**Comment by techwhizbang** [comment]
I am also seeing this issue sporadically in our production clusters. While low volume it is still disruptive and annoying since it causes our API calls to fail and additional boilerplate checks in our client code are needed to retry when this happens.

Can this be re-opened please?

**Comment by ewjmulder** [comment]
Another place where we see this happening: in our Airflow (https://github.com/apache/airflow) workflow there is a point when lots of pods are created at the same time to run some calculations in parallel. From time to time, we hit this bug:

`HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on resourcequotas \"delphi-resourcequota\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"delphi-resourcequota","kind":"resourcequotas"},"code":409}
`

**Comment by CatalinAdlerDF** [comment]
This also happens in our cluster on GKE and we do create job objects at high rate: up to 10k.
We get the job object created but it does not have a pod; pod takes forever to get even get created.

**Comment by MikaelSmith** [comment]
Also see this when installing the GitLab helm chart in CI
```
Release "gitlab" does not exist. Installing it now.
Error: Operation cannot be fulfilled on resourcequotas "gke-resource-quotas": the object has been modified; please apply your changes to the latest version and try again
```
which causes that operation to fail.

**Comment by dprotaso** [comment]
Just encountered this - seems like a lot of clients are building in their own retry logic 🤷‍♂️ 

Unsure if this is something that can be generalized on the API server - ie. when creating objects and you hit a conflict with a dependent resource quota - retry

**Comment by aandryashin** [comment]
IHMO it should be fixed on Kubernetes side, there is issue we have already
created.

ср, 21 апр. 2021 г., 07:24 Dave Protasowski ***@***.***>:

**Comment by aheng-ch** [comment]
When we delete pods, the resource_quota_controller will recalculate the quota and try to update it. When we create pods, the admission_control_plugin (ResourceQuota) will calculate the incoming request and update the observed usage based on the previously read ResourceQuota.ResourceVersion.
If we delete and create pods at the same time, the resource_quota_controller may update the quota unsuccessfully because 
 the admission_control_plugin has updated it. This will cause the quota not to be updated in time. Eg:
1. the quota.used=100
2. delete pods (request 50) and create pods (request 80）at the same time
3. the resource_quota_controller detects pods has been deleted and recalculates the quota.used (50, 100-50)
4. the admission_control_plugin detects pods creating and updates the quota.used to 180 (100+80)
5. the resource_quota_controller updates the quota failed (version expired, the quota.used=180).
6. a few minutes later (resync, 5min), the resource_quota_controller recalculates the quota.used (130) and updates it successfully.

**Comment by amitsingla** [comment]
We are facing the similar issue when we upgraded the Kube to 1.20.15 and moved Kube services to Binary. Is anyone able to find any workaround except the one raised above in helm. ? I am bit hesitant to rollout this in production with this bug as most of our application deployments are helm based and thorough pipelines. It would be great if someone can share and guide us how to fix this or workaround this issue. Thanks
